torch/optim/adadelta.py

```diff
@@ -102,13 +108,111 @@ def step(self, closure=None):
 
                 state['step'] += 1
 
-            F.adadelta(params_with_grad,
-                       grads,
-                       square_avgs,
-                       acc_deltas,
-                       lr=lr,
-                       rho=rho,
-                       eps=eps,
-                       weight_decay=weight_decay)
+            adadelta(params_with_grad,
+                     grads,
+                     square_avgs,
+                     acc_deltas,
+                     lr=lr,
+                     rho=rho,
+                     eps=eps,
+                     weight_decay=weight_decay,
+                     foreach=foreach)
 
         return loss
+
+
+def adadelta(params: List[Tensor],
+             grads: List[Tensor],
+             square_avgs: List[Tensor],
+             acc_deltas: List[Tensor],
+             # kwonly args with defaults are not supported by functions compiled with torchscript issue #70627
+             # setting this as kwarg for now as functional API is compiled by torch/distributed/optim
+             foreach: bool = None,
+             *,
+             lr: float,
+             rho: float,
+             eps: float,
+             weight_decay: float):
+    r"""Functional API that performs Adadelta algorithm computation.
+
+    See :class:`~torch.optim.Adadelta` for details.
+    """
+
+    if foreach is None:
+        # Placeholder for more complex foreach logic to be added when value is not set
+        foreach = False
+
+    if foreach and torch.jit.is_scripting(): 
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if foreach and not torch.jit.is_scripting(): 
```

Comment:
had to leave the not torch.jit.is_scripting() here because I tried many variations of this check and they all raised torchscript error that func is undefined when func is set in the else branch (specific error was `Runtime error: undefined value func`)

variations I tried:
```if foreach and torch.jit.is_scripting(): 
        raise RuntimeError('torch.jit.script not supported with foreach optimizers')

    if foreach: 
        func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```
and 

```
    if foreach:
        if torch.jit.is_scripting():
            raise RuntimeError('torch.jit.script not supported with foreach optimizers')
        else: # with and without else statement
            func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```

and 

```
    if foreach and not torch.jit.is_scripting():
        if torch.jit.is_scripting():
            raise RuntimeError('torch.jit.script not supported with foreach optimizers')
        else: # with and without else statement
            func = _multi_tensor_adadelta
    else:
        func = _single_tensor_adadelta
```

Reply:
note that the `not` is not at the right place in the last one ;)

In any case, the current version is good enough I think